### PR TITLE
Move Llama3 codebase to TT-Transformer

### PR DESCRIPTION
**Pull request by yieldthought**
### Ticket
#18205

### Problem description
The llama3 common codebase has grown up enough to support many different transformer based LLMs (llama, qwen, deepseek distills, Phi3.5, etc.). As such it is time to rename it to what it actually is, an initial version of a TT-Tranformer library.

### What's changed
Move and rename models/demos/llama3 to models/tt-transformers:
- [x] git mv files
- [x] Rename files and includes to remove llama_ prefix
- [x] Rename file imports (careful with reference code)
- [x] Update paths in many files including CI scripts and READMEs
- [x] Rename classes to remove Llama prefix
- [x] Rename env vars
- [x] Rename variables to remove Llama, LLAMA, llama usage
- [x] Rename FAKE_DEVICE to MESH_DEVICE for consistency with vLLM
- [x] Update READMEs
- [x] Check all other metal code for models/demos/llama3
    - [x] Especially CI symlinks and scripts
- [x] Update lt
- [x] Update tech report being careful not to break commit-specific links
- [x] Update CODEOWNERS
- [x] Run accuracy test and demos locally (to check paths etc not broken)
- [x] Run unit tests for one model size locally (to check paths etc not broken)

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/runs/13651624154) CI passes
- [x] [Single card](https://github.com/tenstorrent/tt-metal/actions/runs/13654186124) tests
- [x] [TG](https://github.com/tenstorrent/tt-metal/actions/runs/13654201207) tests - [TG tests fail on main](https://github.com/tenstorrent/tt-metal/issues/18863)
- [x] [Single card again post-rebase](https://github.com/tenstorrent/tt-metal/actions/runs/13764586418)
- [x] [T3K](https://github.com/tenstorrent/tt-metal/actions/runs/13764579466) tests

Tests show some instability but nothing new or related to this PR.

### CODE REVIEW

**Comment by yieldthought** [comment]
Lol of course the link-checker fails, this PR is going to add the destination for the link being checked >.<

**Comment by mtairum** [comment]
FIY @gfengTT @tlevinTT 

Your changes on precision will eventually need to be synced with this PR. Either before or after.

File: models/tt_transformers/tt/generator_vllm.py [hunk 1]
```diff
@@ -99,7 +99,7 @@ def input_processor_for_mllama(ctx: InputContext, inputs: Union[DecoderOnlyInput
     return inputs
 
 
-def input_processor_for_llama_text(ctx: InputContext, inputs: Union[DecoderOnlyInputs, EncoderDecoderInputs]):
+def input_processor_for_text(ctx: InputContext, inputs: Union[DecoderOnlyInputs, EncoderDecoderInputs]):
```
Comment (skhorasganiTT) [inline models/tt_transformers/tt/generator_vllm.py:102]: This function is Llama specific, so it can be kept as `input_processor_for_llama_text`

File: models/tt_transformers/demo/simple_text_demo.py [hunk 2]
```diff
@@ -267,21 +267,21 @@ def create_tt_model(
...
+        ModelOptimizations.accuracy,
     ],
 )
 @pytest.mark.parametrize("device_params", [{"trace_region_size": 23887872, "num_command_queues": 2}], indirect=True)
 @pytest.mark.parametrize(
     "mesh_device",
     [
         {"N150": (1, 1), "N300": (1, 2), "T3K": (1, 8), "TG": (8, 4)}.get(
-            os.environ.get("FAKE_DEVICE"), len(ttnn.get_device_ids())
+            os.environ.get("MESH_DEVICE"), len(ttnn.get_device_ids())
```
Comment (tt-rkim) [inline models/tt_transformers/demo/simple_text_demo.py:279]: We control it via this?

File: models/tt_transformers/README.md [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+
+On N150, N300 and LoudBox / QuietBox you should also set:
+
+```
+export WH_ARCH_YAML=wormhole_b0_80_arch_eth_dispatch.yaml
+```
+
+Description of the environment variables:
+- `HF_MODEL` is the HuggingFace org/name of the model you want to run or the path to the downloaded Huggingface weights.
+- `LLAMA_DIR` sets the path for Meta-provided Llama3 model weights if you are not using HuggingFace.
```
Comment (tt-rkim) [inline models/tt_transformers/README.md:137]: No MESH_DEVICE?

**Review by tt-rkim — APPROVED** [review]

**Comment by tt-rkim** [comment]
Looks good, script changes should be ok... I'd recommend at least one demo pipeline

File: models/tt_transformers/demo/simple_text_demo.py (hunk 2 above)
Comment (yieldthought) [reply models/tt_transformers/demo/simple_text_demo.py:279]: The default is to use the entire device for tensor-parallelism, but this lets devs and users run on a smaller number of wormholes on a larger device, v. useful when testing changes on a T3K or TG for example.

File: models/tt_transformers/README.md (hunk 3 above)
Comment (yieldthought) [reply models/tt_transformers/README.md:137]: Was addressed further down the README. Moved up to this section for clarity 🙏

File: models/tt_transformers/demo/simple_text_demo.py [hunk 4]
```diff
@@ -20,12 +20,12 @@
...
-from models.demos.llama3.tt.generator import LlamaGenerator
-from models.demos.llama3.tt.model_config import LlamaOptimizations
-from models.demos.llama3.tt.llama_common import (
+from models.tt_transformers.tt.generator import Generator
+from models.tt_transformers.tt.model_config import ModelOptimizations
+from models.tt_transformers.tt.common import (
     preprocess_inputs_prefill,
     get_rot_transformation_mat,
-    encode_prompt_llama_instruct,
+    encode_prompt_instruct,
```
Comment (cglagovichTT) [inline models/tt_transformers/demo/simple_text_demo.py:28]: some functions may be Llama-specific, like prompt encoding. probably not an issue now but something we should be ready to support as new models have different prompt schemes

File: models/tt_transformers/scripts/repack_weights.py [hunk 5]
```diff
@@ -4,7 +4,7 @@
 
 import torch
 import argparse
-from models.demos.t3000.mixtral8x7b.tt.model_config import TtModelArgs
+from models.demos.t3000.mixtral8x7b.tt.model_config import ModelArgs
```
Comment (cglagovichTT) [inline models/tt_transformers/scripts/repack_weights.py:7]: This script is Meta-Llama specific IIRC

File: models/tt_transformers/tests/test_chunked_generation.py [hunk 6]
```diff
@@ -7,14 +7,14 @@
...
     get_block_size,
     num_blocks_in_seq,
 )
-from models.demos.llama3.tt.llama_model import TtTransformer
-from models.demos.llama3.tt.model_config import TtModelArgs, LlamaOptimizations
-from models.demos.llama3.tt.generator import LlamaGenerator
+from models.tt_transformers.tt.model import Transformer
+from models.tt_transformers.tt.model_config import ModelArgs, ModelOptimizations
+from models.tt_transformers.tt.generator import Generator
 from models.demos.t3000.llama2_70b.reference.llama.llama31_8b.model import Transformer
```
Comment (cglagovichTT) [inline models/tt_transformers/tests/test_chunked_generation.py:18]: problem that we import `Transformer` from two different modules?

**Review by cglagovichTT — APPROVED** [review]
some comments. nice :)

File: README.md [hunk 7]
```diff
@@ -29,21 +29,21 @@
...
-| [Llama 3.1 70B (TP=32)](./models/demos/llama3)                | 32   | [Galaxy](https://tenstorrent.com/hardware/galaxy)        | 763       | 13.5  | 80              | 432.0 | [v0.56.0-rc6](https://github.com/tenstorrent/tt-metal/tree/v0.56.0-rc6) | [b9564bf](https://github.com/tenstorrent/vllm/tree/b9564bf364e95a3850619fc7b2ed968cc71e30b7) |
-
+| [Llama 3.1 70B (TP=32)](./models/tt_transformers)                | 32   | [Galaxy](https://tenstorrent.com/hardware/galaxy)        | 763       | 13.5  | 80              | 432.0 | [v0.56.0-rc6](https://github.com/tenstorrent/tt-metal/tree/v0.56.0-rc6) | [b9564bf](https://github.com/tenstorrent/vllm/tree/b9564bf364e95a3850619fc7b2ed968cc71e30b7) |
+| [DeepSeek R1 Distill Llama 3.3 70B (TP=8)](./models/tt_transformers)       | 32    | [QuietBox](https://tenstorrent.com/hardware/tt-quietbox) | 1113       | 16.4  | 33              |524.8 | [main](https://github.com/tenstorrent/tt-metal/) | [b9564bf](https://github.com/tenstorrent/vllm/tree/b9564bf364e95a3850619fc7b2ed968cc71e30b7)      |
```
Comment (skhorasganiTT) [inline README.md:46]: I think this Deepseek row was re-introduced by mistake, the correct row already exists above

File: models/tt_transformers/demo/simple_text_demo.py (hunk 4 above)
Comment (yieldthought) [reply models/tt_transformers/demo/simple_text_demo.py:28]: Already done, `encode_prompt_instruct` is not used and instead model_config.encode_promompt calls the appropriate huggingface prompt templating code if it's a hf model 🙌 
Removing this dead code.

File: README.md (hunk 7 above)
Comment (yieldthought) [reply README.md:46]: Nice catch, thanks.

File: models/tt_transformers/tests/test_chunked_generation.py (hunk 6 above)
Comment (yieldthought) [reply models/tt_transformers/tests/test_chunked_generation.py]: Nice catch, fixed and test passes

**Review by skhorasganiTT — APPROVED** [review]

File: models/tt_transformers/tt/multimodal/llama_cross_attention_transformer_text.py [hunk 8]
```diff
@@ -6,14 +6,14 @@
 import ttnn
 import torch
 from tqdm import tqdm
-from models.demos.llama3.tt.llama_decoder import TtTransformerBlock
-from models.demos.llama3.tt.multimodal.llama_cross_block import TtLlamaCrossAttentionTransformerBlock
-from models.demos.llama3.tt.distributed_norm import DistributedNorm
+from models.tt_transformers.tt.decoder import TtTransformerBlock
```
Comment (cglagovichTT) [inline models/tt_transformers/tt/multimodal/llama_cross_attention_transformer_text.py:9]: Few more instances of TtTransformerBlock (and other Tt prefix imports) are left, causing import errors on `simple_vision_demo`

**Review by uaydonat — APPROVED** [review]

**Review by mtairum — APPROVED** [review]
This is a huge milestone to what has started as a common codebase for Llama3, now slowly being transformed into a proper TT-Transformers library.

🙌 🙌 🙌
